mlflow/pytorch/_pytorch_autolog.py

```diff
@@ -0,0 +1,235 @@
+            try_mlflow_log(mlflow.set_tag, "Mode", "training")
+            try_mlflow_log(mlflow.log_param, "epochs", trainer.max_epochs)
+
+            for callback in trainer.callbacks:
+                if isinstance(callback, pl.callbacks.early_stopping.EarlyStopping):
+                    self.early_stopping = True
+                    self._log_early_stop_params(callback)
+
+            if hasattr(trainer, "optimizers"):
+                for optimizer in trainer.optimizers:
+                    try_mlflow_log(mlflow.log_param, "optimizer_name", type(optimizer).__name__)
+                    optimizer_name = type(optimizer).__name__.lower() + "_optimizer"
+
+                    if hasattr(optimizer, "defaults"):
+                        optim_dict = optimizer.defaults
+
+                        if "lr" in optim_dict:
+                            try_mlflow_log(
+                                mlflow.log_param,
+                                "learning_rate_" + optimizer_name,
+                                optim_dict["lr"],
+                            )
+
+                        if "eps" in optim_dict:
+                            try_mlflow_log(
+                                mlflow.log_param, "epsilon_" + optimizer_name, optim_dict["eps"]
+                            )
+
+                        if "betas" in optim_dict:
+                            try_mlflow_log(
+                                mlflow.log_param, "betas_" + optimizer_name, optim_dict["betas"]
+                            )
+
+                        if "weight_decay" in optim_dict:
+                            try_mlflow_log(
+                                mlflow.log_param,
+                                "weight_decay_" + optimizer_name,
+                                optim_dict["weight_decay"],
+                            )
+
+            summary = str(ModelSummary(pl_module, mode="full"))
+            tempdir = tempfile.mkdtemp()
+            try:
+                summary_file = os.path.join(tempdir, "model_summary.txt")
+                with open(summary_file, "w") as f:
+                    f.write(summary)
+
+                try_mlflow_log(mlflow.log_artifact, local_path=summary_file)
+            finally:
+                shutil.rmtree(tempdir)
+
+        def on_train_end(self, trainer, pl_module):
+            """
+            Logs the model checkpoint into mlflow - models folder on the training end
+
+            :param trainer: pytorch lightning trainer instance
+            :param pl_module: pytorch lightning base module
+            """
+
+            mlflow.pytorch.log_model(pytorch_model=trainer.model, artifact_path="model")
+
+            if self.early_stopping and trainer.checkpoint_callback.best_model_path:
+                try_mlflow_log(
+                    mlflow.log_artifact,
+                    local_path=trainer.checkpoint_callback.best_model_path,
+                    artifact_path="restored_model_checkpoint",
+                )
```

Comment:
Hi @shrinath-suresh,

When training is early stopped, does `trainer` restore the best weights?

Reply:
Looks like it doesn't:

https://github.com/PyTorchLightning/pytorch-lightning/blob/4018237c309b7d9d6978da73132003615341e04a/pytorch_lightning/trainer/training_loop.py#L591

Reply:
Not entirely sure what the question is, when does the trainer restore the model's weights?

Reply:
@SeanNaren Thanks for the reply.

```md
Early Stopping patience: 3

| Epoch | Validation loss | comment                |
| :---- | :-------------- | ---------------------- |
| 1     | 9               |                        |
| 2     | 8               |                        |
| 3     | 7               |                        | <- best
| 4     | 7.1             | loss starts increasing |
| 5     | 7.2             |                        |
| 6     | 7.3             | training early-stops   | <- stopped
| 7     | -               |                        |
| 8     | -               |                        |
| 9     | -               |                        |
```

Reply:
Keras' early-stopping has this featurs (`restore_best_weights`):
https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/EarlyStopping

Reply:
Found a related discussion: https://github.com/PyTorchLightning/pytorch-lightning/issues/1395

Reply:
Hi, @harupy , it seems that until now mlflow still doesn't restore the best weights when training is early stopped, are there any updates on this feature?